Modules/_decimal/_decimal.c

```diff
@@ -2323,38 +2322,46 @@ static PyObject *
 dec_from_long(decimal_state *state, PyTypeObject *type, PyObject *v,
               const mpd_context_t *ctx, uint32_t *status)
 {
-    PyObject *dec;
-    PyLongObject *l = (PyLongObject *)v;
+    PyObject *dec = PyDecType_New(state, type);
 
-    dec = PyDecType_New(state, type);
     if (dec == NULL) {
         return NULL;
     }
 
-    if (_PyLong_IsZero(l)) {
-        _dec_settriple(dec, MPD_POS, 0, 0);
-        return dec;
-    }
-
-    uint8_t sign = _PyLong_IsNegative(l) ? MPD_NEG :  MPD_POS;
+    PyLongExport export_long;
 
-    if (_PyLong_IsCompact(l)) {
-        _dec_settriple(dec, sign, l->long_value.ob_digit[0], 0);
-        mpd_qfinalize(MPD(dec), ctx, status);
-        return dec;
+    if (PyLong_Export(v, &export_long) == -1) {
+        Py_DECREF(dec);
+        return NULL;
     }
-    size_t len = _PyLong_DigitCount(l);
+    if (export_long.digits) {
+        const PyLongLayout *layout = PyLong_GetNativeLayout();
+        const uint32_t base = (uint32_t)1 << layout->bits_per_digit;
+        const uint8_t sign = export_long.negative ? MPD_NEG : MPD_POS;
+        const Py_ssize_t len = export_long.ndigits;
 
-#if PYLONG_BITS_IN_DIGIT == 30
-    mpd_qimport_u32(MPD(dec), l->long_value.ob_digit, len, sign, PyLong_BASE,
-                    ctx, status);
-#elif PYLONG_BITS_IN_DIGIT == 15
-    mpd_qimport_u16(MPD(dec), l->long_value.ob_digit, len, sign, PyLong_BASE,
-                    ctx, status);
-#else
-  #error "PYLONG_BITS_IN_DIGIT should be 15 or 30"
-#endif
+        if (base > UINT16_MAX) {
+            mpd_qimport_u32(MPD(dec), export_long.digits, len, sign,
+                            base, ctx, status);
+        }
+        else {
+            mpd_qimport_u16(MPD(dec), export_long.digits, len, sign,
+                            base, ctx, status);
+        }
+        PyLong_FreeExport(&export_long);
+    }
+    else {
+        const int64_t value = export_long.value;
 
+        if (-(int64_t)UINT32_MAX <= value && value <= (int64_t)UINT32_MAX) {
+            _dec_settriple(dec, value < 0 ? MPD_NEG : MPD_POS,
+                           (uint32_t)Py_ABS(value), 0);
+            mpd_qfinalize(MPD(dec), ctx, status);
```

Comment:
Is it worth it to have this fast-path? (Why not always calling mpd_qset_i64()?)

Reply:
It's up to reviewers.  In old pr I did benchmarks:
https://github.com/python/cpython/pull/127925#issuecomment-2546186276